### PR TITLE
New openstack create and delete roles and playbooks

### LINKED ISSUE #210: Improve the openstack playbook to support to use passwordstore

### PR DESCRIPTION
**Pull request by jacobdotcosta**
These new roles al playbooks use the passwordstore.

Closes #210

### CODE REVIEW

File: ansible/roles/openstack/create_vm/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    - include_role:
+        name: generate_inventory
+      vars:
+        ip_address: "{{ openstack_vm_ipv4  }}"
+        username: "{{ login_username }}"
+        # keyfile: "{{ private_key_in_inventory }}"
+        keyfile: "{{ key_path }}"
+        hostname: "{{ openstack.vm.name }}"
+        openshift_origin_version: "3.11"
```
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply ansible/roles/openstack/create_vm/tasks/main.yml]: fixed

File: ansible/roles/openstack/create_vm/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  vm:
+    title: CentOS 7 Generic Cloud Latest
+    name: "{{ hostname }}"
+    image: CentOS-7-x86_64-GenericCloud-released-latest
+    key_name: "{{ hostname }}"
+    flavor: m1.xlarge
+    network: default_network
+    security_group: default_security_group
+    volumes: []
```
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply ansible/roles/openstack/create_vm/defaults/main.yml]: fixed

File: ansible/roles/openstack/create_vm/tasks/keys.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  os_keypair:
+    state: absent
+    auth:
+      project_name: "{{ openstack.os_project_name }}"
+      # username: "{{ openstack.os_username }}"
+      # password: "{{ openstack.os_password }}"
+      username: "{{ openstack_console_user }}"
+      password: "{{ openstack_console_password }}"
+      # user_domain_name: "{{ openstack.os_domain }}"
+      # project_domain_name: "{{ openstack.os_domain }}"
```
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply ansible/roles/openstack/create_vm/tasks/keys.yml]: comments removed

File: ansible/roles/openstack/create_vm/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+openstack:
+  os_project_name: spring-boot-jenkins
+  
+  sshkey_name: "{{ hostname }}"
+
+  vm:
+    title: CentOS 7 Generic Cloud Latest
+    name: "{{ hostname }}"
+    image: CentOS-7-x86_64-GenericCloud-released-latest
```
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply ansible/roles/openstack/create_vm/defaults/main.yml]: Created #239 for this.

File: ansible/roles/openstack/create_vm/tasks/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobdotcosta) [reply ansible/roles/openstack/create_vm/tasks/main.yml]: No, that variable is not being used in any way.

File: ansible/playbook/openstack_vm_create.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+- name:  "Wait for the VM to boot and we can ssh"
+  hosts: "{{ vm_name }}"
+  gather_facts: no
+
+  tasks:
+    - name: "Show 'Wait for connection to host' output"
+      debug:
+        msg: 
+          - "ip : {{ query('passwordstore', '/openstack/' + vm_name + '/ansible_ssh_host')[0] }}"
```
Comment (cmoulliard) [inline ansible/playbook/openstack_vm_create.yml:37]: Is there a control/check verifying if password store is installed on the users machine ?

File: ansible/playbook/openstack_vm_create_aggregate.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  import_playbook: "passstore_controller_inventory.yml"
+  vars:
+    vm_user: "centos"
+    pass_provider: "openstack"
+  tags: [create]
+
+- name: "Create Server on Openstack"
+  import_playbook: "openstack_vm_create.yml"
+  vars:
+    hetzner_context_name: "snowdrop"
```
Comment (cmoulliard) [inline ansible/playbook/openstack_vm_create_aggregate.yml:22]: Why do we have a var `hetzner_context_name` part of this playbook ?

File: ansible/playbook/openstack_vm_remove_aggregate.yml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+---
+# Variables:
+#  - "vm_name": "n311-test"
+- name: "Delete Server on Openstack"
+  import_playbook: "openstack_vm_delete.yml"
+  vars:
+    hetzner_context_name: "snowdrop"
```
Comment (cmoulliard) [inline ansible/playbook/openstack_vm_remove_aggregate.yml:7]: Same remark about the var as mentioned before !

File: openstack/README.md [hunk 8]
```diff
@@ -1,38 +1,133 @@
...
 
+This can of course be done via the OpenStack UI or can be automated using our Ansible openstack playbook. For the sake of the examples, and since the VM name will be used around several steps, the `VM_NAME` env variable will store the vm name.
+
+```bash
+$ VM_NAME=vm20210221-t01
 ```
-ansible-playbook playbook/openstack.yml -e type=cloude -e '{"state": "present", "hostname": "somehostname", "openstack": {"os_username": "username", "os_password": "password", "os_domain": "domain", "os_auth_url": "https://somehost:13000/v3", "os_project_id": "someprojectid", "os_network": "network", "os_security_group": "security_group"}}'`
+
+```bash
+$ ansible-playbook playbook/openstack_vm_create_aggregate.yml -e k8s_type=masters -e k8s_version=121 -e '{"openstack": {"vm": {"network": "provider_net_shared"}}}' -e vm_name=${VM_NAME} --tags create
```
Comment (cmoulliard) [inline openstack/README.md:38]: Why do you use this bizarre wording as I mentioned on the Hetzner PR `openstack_vm_create_aggregate.yml` and not something like `openstack_vm_k8s.yml` or `openstack_vm_k8s_provisioning.yml` ?

File: ansible/playbook/openstack_vm_create.yml (hunk 5 above)
Comment (jacobdotcosta) [reply ansible/playbook/openstack_vm_create.yml:50]: I have implemented a new playbook to validate our passwordstore when creating openstack VMs. Will later apply to the hetzner VMs when both PR are merged.

File: ansible/playbook/openstack_vm_create_aggregate.yml (hunk 6 above)
Comment (jacobdotcosta) [reply ansible/playbook/openstack_vm_create_aggregate.yml:26]: removed this reference as it wasn't being used

File: ansible/playbook/openstack_vm_remove_aggregate.yml (hunk 7 above)
Comment (jacobdotcosta) [reply ansible/playbook/openstack_vm_remove_aggregate.yml]: the hetzner variable definition has been removed as it wasn't used.

File: openstack/README.md (hunk 8 above)
Comment (jacobdotcosta) [reply openstack/README.md:38]: The idea of these aggregate playbooks is of having a higher level of abstraction reusing existing playbooks. For instance, in order to create a openstack VM we have to execute several playbooks so the aggregate playbook calls all of them.  In this case it calls the following playbooks:
* passstore_controller_check.yml
* passstore_controller_inventory.yml
* openstack_vm_create.yml
* sec_host.yml
for now they don't deploy k8s.

I can call them differently but I wanted to separate them from the "_unitary_" playbooks

Comment (cmoulliard) [reply openstack/README.md:38]: Keep then the name. The most important is to be sure that when someone will execute the command they will not have to write supeeeeeer long command name. We should perhaps create within the ansible-k8s tooling  image some aliases to help the user to create a vm+k8s or delete a vm+k8s

Comment (jacobdotcosta) [reply openstack/README.md:38]: I understand, I'll try finding a better name

Comment (jacobdotcosta) [reply openstack/README.md:43]: I'd like that we manage the k8s on another PR as we also have to adapt the installation for new k8s versions.

**Pull request merged**